sae_lens/saes/sae_base.py

```diff
@@ -0,0 +1,588 @@
+            "seqpos_slice": self.seqpos_slice,
+            "neuronpedia_id": self.neuronpedia_id,
+        }
+
+
+class BaseTrainingSAE(BaseSAE, ABC):
+    """Abstract base class for training versions of SAEs."""
+
+    cfg: "TrainingSAEConfig"  # Quote marks make this a forward reference
+
+    def __init__(self, cfg: TrainingSAEConfig, use_error_term: bool = False):
+        super().__init__(cfg, use_error_term)
+
+        # Turn off hook_z reshaping for training mode - the activation store
+        # is expected to handle reshaping before passing data to the SAE
+        self.turn_off_forward_pass_hook_z_reshaping()
+
+        self.mse_loss_fn = self._get_mse_loss_fn()
+
+    @abstractmethod
+    def encode_with_hidden_pre(
+        self, x: Float[torch.Tensor, "... d_in"]
+    ) -> Tuple[Float[torch.Tensor, "... d_sae"], Float[torch.Tensor, "... d_sae"]]:
+        """Encode with access to pre-activation values for training."""
+        pass
+
+    @abstractmethod
+    def calculate_aux_loss(
+        self,
+        feature_acts: torch.Tensor,
+        hidden_pre: torch.Tensor,
+        dead_neuron_mask: Optional[torch.Tensor] = None,
+        current_l1_coefficient: Optional[float] = None,
+        sae_in: Optional[torch.Tensor] = None,
+        sae_out: Optional[torch.Tensor] = None,
+        **kwargs: Any,
+    ) -> torch.Tensor:
+        """Calculate architecture-specific auxiliary loss terms."""
+        pass
+
+    def training_forward_pass(
+        self,
+        sae_in: torch.Tensor,
+        current_l1_coefficient: float,
+        dead_neuron_mask: Optional[torch.Tensor] = None,
+    ) -> TrainStepOutput:
+        """Forward pass during training."""
+        feature_acts, hidden_pre = self.encode_with_hidden_pre(sae_in)
+        sae_out = self.decode(feature_acts)
+
+        # Calculate MSE loss
+        per_item_mse_loss = self.mse_loss_fn(sae_out, sae_in)
+        mse_loss = per_item_mse_loss.sum(dim=-1).mean()
+
+        # Calculate architecture-specific auxiliary losses
+        aux_losses = self.calculate_aux_loss(
+            feature_acts=feature_acts,
+            hidden_pre=hidden_pre,
+            dead_neuron_mask=dead_neuron_mask,
+            current_l1_coefficient=current_l1_coefficient,
+            sae_in=sae_in,
+            sae_out=sae_out,
+        )
+
+        # Total loss is MSE plus all auxiliary losses
+        total_loss = mse_loss
+
+        # Create losses dictionary with mse_loss
+        losses = {"mse_loss": mse_loss}
+
+        # Add architecture-specific losses to the dictionary
+        # Make sure aux_losses is a dictionary with string keys and tensor values
+        if isinstance(aux_losses, dict):
+            losses.update(aux_losses)
+
+        # Sum all losses for total_loss
+        if isinstance(aux_losses, dict):
+            for loss_value in aux_losses.values():
+                total_loss = total_loss + loss_value
```

Comment:
in this case `aux_losses` should be merged into `losses` so they all get reported in wandb